File: NewsAlarm/app/src/main/java/com/alarm/newsalarm/utils/LogUtil.java
package com.alarm.newsalarm.utils;

import android.util.Log;

import androidx.annotation.NonNull;

public class LogUtil {

    private enum LogLevel {
        DEBUG, INFO, WARN, ERROR
    }

    private static final String TAG = "NewsAlarm";

    private LogUtil() {
    }

    private static void log(@NonNull String className, @NonNull String methodName,
            @NonNull String message, @NonNull LogLevel logLevel) {

        String logMsg = "[" + className + "$" + methodName + "] " + message;

        switch (logLevel) {
            case DEBUG -> Log.d(TAG, logMsg);
            case INFO -> Log.i(TAG, logMsg);
            case WARN -> Log.w(TAG, logMsg);
            case ERROR -> Log.e(TAG, logMsg);
        }
    }

    public static void logD(String className, String methodName, String message) {
        log(className, methodName, message, LogLevel.DEBUG);
    }

    public static void logI(String className, String methodName, String message) {
        log(className, methodName, message, LogLevel.INFO);
    }

    public static void logW(String className, String methodName, String message) {
        log(className, methodName, message, LogLevel.WARN);
    }

    public static void logE(String className, String methodName, String message) {
        log(className, methodName, message, LogLevel.ERROR);
    }

    public static void logE(String className, String methodName, Exception e) {
        StackTraceElement[] stackTrace = e.getStackTrace();
        for (StackTraceElement element : stackTrace) {
            log(className, methodName, element.toString(), LogLevel.ERROR);
        }
    }
}

File: NewsAlarmKt/app/src/main/java/com/alarm/newsalarmkt/utils/LogUtil.kt
package com.alarm.newsalarmkt.utils

import android.util.Log

object LogUtil {
    private const val TAG = "NewsAlarm"

    private fun log(
        className: String, methodName: String,
        message: String, logLevel: LogLevel
    ) {
        val logMsg = "[$className$$methodName] $message"

        when (logLevel) {
            LogLevel.DEBUG -> Log.d(TAG, logMsg)
            LogLevel.INFO -> Log.i(TAG, logMsg)
            LogLevel.WARN -> Log.w(TAG, logMsg)
            LogLevel.ERROR -> Log.e(TAG, logMsg)
        }
    }

    @JvmStatic
    fun logD(className: String, methodName: String, message: String) {
        log(className, methodName, message, LogLevel.DEBUG)
    }

    @JvmStatic
    fun logI(className: String, methodName: String, message: String) {
        log(className, methodName, message, LogLevel.INFO)
    }

    @JvmStatic
    fun logW(className: String, methodName: String, message: String) {
        log(className, methodName, message, LogLevel.WARN)
    }

    @JvmStatic
    fun logE(className: String, methodName: String, message: String) {
        log(className, methodName, message, LogLevel.ERROR)
    }

    @JvmStatic
    fun logE(className: String, methodName: String, e: Exception) {
        val stackTrace = e.stackTrace
        for (element in stackTrace) {
            log(className, methodName, element.toString(), LogLevel.ERROR)
        }
    }

    private enum class LogLevel {
        DEBUG, INFO, WARN, ERROR
    }
}
